### DISCUSSION TITLE
GetAnchors() on MediaField query

### DISCUSSION THREAD

**Discussion by scleaver** [OrchardCMS/OrchardCore#8616] [Q&A]
I have a strongly type part called `MyPart` which has a property of type `MediaField` with the name `Image`. I attached an image and set an anchor point to the first image. In code this does not seem to work (assuming contentItem is an content item):

```
var myPart = contentItem.As<MyPart>();
var mediaAnchor = new Anchor();
var anchors = myPart.Image.GetAnchors();
if(anchors !=null)
{
    mediaAnchor = anchors[0];
}
```
A `system.IndexOutOfRangeException: Index was outside the bounds of the array.` is thrown. While the GetAnchors is not playing nicely the `myPart.Image.Paths[0]` works fine though.

However if my code is like this it is fine:
```
var mediaAnchor = new Anchor();
var anchors = (Anchor[])contentItem.ContentItem.Content.GroupPart.Image.Anchors.ToObject<Anchor[]>();
if(anchors != null)
{
    mediaAnchor = anchors[0];
}
```
Am I using GetAnchors incorrectly or should I raise an issue?

**Comment by deanmarcussen** [comment]
Yes this is known behavior but we have not addressed it at this point.

because we store the anchors on the Content property of the field the .As conversion does not take this into account and the data is lost during conversion.

@netwavebe and I discussed this a while ago but we never reached a good conclusion and I have forgotten some of the options we talked about.

there might be something we can do in the .As conversion to bring by these through or we might use some hard typed models to assist the conversion process

**Comment by netwavebe** [comment]
Like @deanmarcussen said: the As<T> conversion is where the anchor information is lost.

I solved it by creating a 'extended' media field like this:

```
public class AnchoredMediaField : MediaField
{
	public Anchor[] Anchors { get; set; }
}
```

Change the type of `Image` from `MediaField` to `AnchoredMediaField` in class `MyPart`.

You can now use `As<MyPart>().Image.Anchors`. Hope this helps.

**Comment by netwavebe** [comment]
In case it might help you, this is the complete class I'm using. It allows me to get image information together with their (optional) anchors and mediatext:

```
public class AnchoredMediaField : MediaField
{
	public Anchor[] Anchors { get; set; }

	public ImageItem[] Items
	{
		get
		{
			var items = new List<ImageItem>();

			foreach (var path in Paths)
			{
				var index = Array.IndexOf(Paths, path);

				var mediaText = MediaTexts != null && index >= 0 && index < MediaTexts.Length ? MediaTexts[index] : null;
				var anchor = Anchors != null && index >= 0 && index < Anchors.Length ? Anchors[index] : null;

				items.Add(new ImageItem(path, mediaText, anchor));
			}

			return items.ToArray();
		}
	}
}

public class ImageItem
{
	public string Path { get; set; }
	public string MediaText { get; set; }
	public Anchor Anchor { get; set; }

	public ImageItem(string path, string mediaText, Anchor anchor)
	{
		Path = path;
		MediaText = mediaText;
		Anchor = anchor;
	}
}
```

**Reply by scleaver** [reply]
Nice one @netwavebe - I am feeling a similar approach should be baked in. @deanmarcussen - how do you feel about this?

**Reply by deanmarcussen** [reply]
Yes I agree we should bake something in, and this looks good.

The only thing I'd want to check is what happens when you call `BuildDisplay` on a content item that has this different media field in.

The display drivers may or may not get confused about the concrete / inherited type. They're not especially built for inheritance, so might (or might not) result in odd behaviour.

But @scleaver feel free to open as an issue and try it to see

**Reply by scleaver** [reply]
Couldn't we implement the same in a MediaField rather than extend it with AnchoredMediaField?
